Interchain exchange tutorial (#573)
cosmos/sdk-tutorials: interchain-exchange/tutorial/01-app-design.md

---
order: 2
---

# App Design

In this chapter you will learn how the interchain exchange module is designed.
The module has orderbooks, buy- and sell orders.
First, an orderbook for a pair of token has to be created.

> After an ordebrook exists, you can create buy and sell orders for this pair of token.

Reviewer: ```suggestion
After an orderbook exists, you can create buy and sell orders for this pair of token.
```

This process can be reversed when the `voucher` get burned again to enable back the original token. This will be explained throghout the tutorial in more detail.

## Assumption

For any pair of tokens you can create an orderbook. The orderbook can be created for the exchange of any tokens between any pair of chains. The requirement is to have the `ibcdex` module available. There can only be one orderbook for a pair of token at the same time.
<!-- There is no condition to check for open channels between two chains. -->
A specific chain cannot mint new of its native token. 
<!-- The module is trustless, there is no condition to check when opening a channel between two chains. Any pair of tokens can be exchanged between any pair of chains. -->

This module is inspired by the [`ibc-transfer`](https://github.com/cosmos/cosmos-sdk/tree/v0.42.1/x/ibc/applications/transfer) module and will have some similarities, like the `voucher` creation. It will be more complex but it will display how to create:

- Several types of packets to send
- Several types of acknowledgments to treat
- Some more complex logic on how to treat a packet, on receipt, on timeout and more

## Overview

> Assume you have two blockchains, one on `Venus`, one on `Mars`. The token to use on Venus is called `vcx`, the token on Mars `mcx`. 

Reviewer: ```suggestion
Assume you have two blockchains: `Venus` and `Mars`. The native token on Venus is called `vcx`, the token on Mars is `mcx`. 
```

> When exchanging a token from Mars to Venus, on the Venus blockchain you would end up with an IBC `voucher` token that looks like `ibc/Venus/mcx`. The IBC path indicates the path of the token. 

Reviewer: ```suggestion
When exchanging a token from Mars to Venus, on the Venus blockchain you would end up with an IBC `voucher` token with a denom that looks like `ibc/B5CB286...A7B21307F `. The long string of characters after `ibc/` is a denom trace hash of a token transferred through IBC. Using the blockchain's API you can get a denom trace from that hash. Denom trace consists of a `base_denom` and a `path`. In our example, `base_denom` will be `mcx` and the `path` will contain pairs of ports and channels through which the token has been transferred. For a single-hop transfer `path` will look like `transfer/channel-0`. Learn more about token paths in [ICS 20](https://github.com/cosmos/ics/tree/master/spec/ics-020-fungible-token-transfer).
```
